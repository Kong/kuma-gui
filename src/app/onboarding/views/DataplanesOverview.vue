<template>
  <OnboardingPage>
    <template #header>
      <OnboardingHeading
        :title="title"
        :description="description"
      />
    </template>

    <template #content>
      <div
        v-if="!tableData.data.length"
        class="justify-center flex my-4"
      >
        <LoadingBox />
      </div>

      <div v-else>
        <div class="flex justify-center mt-10 mb-16 pb-16">
          <div class="w-full sm:w-3/5 p-4">
            <p class="font-bold mb-4">
              Found {{ tableData.data.length }} DPPs:
            </p>

            <KTable
              class="onboarding-dataplane-table"
              :fetcher="() => tableData"
              :headers="tableHeaders"
              disable-pagination
              is-small
            >
              <template #status="{ rowValue }">
                <div
                  class="entity-status"
                  :class="{
                    'is-offline': rowValue.toLowerCase() === 'offline' || rowValue === false,
                    'is-online': rowValue.toLowerCase() === 'online',
                    'is-degraded': rowValue.toLowerCase() === 'partially degraded',
                    'is-not-available': rowValue.toLowerCase() === 'not available',
                  }"
                >
                  <span>{{ rowValue }}</span>
                </div>
              </template>
            </KTable>
          </div>
        </div>
      </div>
    </template>

    <template #navigation>
      <OnboardingNavigation
        next-step="onboarding-completed"
        previous-step="onboarding-add-services-code"
        :should-allow-next="tableData.data.length > 0"
      />
    </template>
  </OnboardingPage>
</template>

<script>
import { KTable } from '@kong/kongponents'

import { PRODUCT_NAME, OFFLINE } from '@/constants'
import { getItemStatusFromInsight } from '@/utilities/dataplane'
import { kumaApi } from '@/api/kumaApi'
import LoadingBox from '@/app/common/LoadingBox.vue'
import OnboardingNavigation from '../components/OnboardingNavigation.vue'
import OnboardingHeading from '../components/OnboardingHeading.vue'
import OnboardingPage from '../components/OnboardingPage.vue'

export default {
  name: 'DataplanesOverview',
  components: {
    OnboardingNavigation,
    OnboardingHeading,
    OnboardingPage,
    LoadingBox,
    KTable,
  },
  metaInfo() {
    return {
      title: this.title,
    }
  },
  data() {
    return {
      productName: PRODUCT_NAME,
      tableHeaders: [
        { label: 'Mesh', key: 'mesh' },
        { label: 'Name', key: 'name' },
        { label: 'Status', key: 'status' },
      ],
      tableData: {
        total: 0,
        data: [],
      },
      timeout: null,
    }
  },

  computed: {
    title() {
      if (this.tableData.data.length) {
        return 'Success'
      }

      return 'Waiting for DPPs'
    },
    description() {
      if (this.tableData.data.length) {
        return 'The following data plane proxies (DPPs) are connected to the control plane:'
      }

      return null
    },
  },
  created() {
    this.getAllDataplanes()
  },
  beforeUnmount() {
    clearTimeout(this.timeout)
  },
  methods: {
    async getAllDataplanes() {
      let shouldRefetch = false
      const result = []

      try {
        const { items } = await kumaApi.getAllDataplanes({ size: 10 })

        if (Array.isArray(items)) {
          for (let i = 0; i < items.length; i++) {
            const { name, mesh } = items[i]

            const { status } = await kumaApi.getDataplaneOverviewFromMesh({ mesh, name }).then((response) =>
              getItemStatusFromInsight(response.dataplaneInsight),
            )

            if (status === OFFLINE) {
              shouldRefetch = true
            }

            result.push({
              status,
              name,
              mesh,
            })
          }
        }
      } catch (e) {
        console.error(e)
      }

      this.tableData.data = result
      this.tableData.total = this.tableData.data.length

      if (shouldRefetch) {
        this.timeout = setTimeout(() => {
          this.getAllDataplanes()
        }, 1000)
      }
    },
  },
}
</script>

<style lang="scss">
.onboarding-dataplane-table tbody tr {
  // hack to do not allow shadow overflow on top/bottom
  box-shadow: -3px 0 0 -1px var(--KTableBorder, var(--steel-200, #dae3f2)) !important;
}
</style>
