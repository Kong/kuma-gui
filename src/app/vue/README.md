---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

