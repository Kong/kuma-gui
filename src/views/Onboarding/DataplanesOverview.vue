<template>
  <OnboardingPage>
    <template #header>
      <OnboardingHeading
        :title="title"
        :description="description"
      />
    </template>
    <template #content>
      <div
        v-if="!tableData.data.length"
        class="justify-center flex my-4"
      >
        <LoadingBox />
      </div>
      <div v-else>
        <div class="flex justify-center mt-10 mb-16 pb-16">
          <div class="w-full sm:w-3/5 p-4">
            <p class="font-bold mb-4">
              Found {{ tableData.data.length }} DPPs:
            </p>
            <KTable
              class="onboarding-dataplane-table"
              :fetcher="() => tableData"
              :headers="tableHeaders"
              disable-pagination
              is-small
            >
              <template #status="{ rowValue }">
                <div
                  class="entity-status"
                  :class="{
                    'is-offline': rowValue.toLowerCase() === 'offline' || rowValue === false,
                    'is-online': rowValue.toLowerCase() === 'online',
                    'is-degraded': rowValue.toLowerCase() === 'partially degraded',
                    'is-not-available': rowValue.toLowerCase() === 'not available',
                  }"
                >
                  <span>{{ rowValue }}</span>
                </div>
              </template>
            </KTable>
          </div>
        </div>
      </div>
    </template>

    <template #navigation>
      <OnboardingNavigation
        next-step="onboarding-completed"
        previous-step="onboarding-add-services-code"
        :should-allow-next="tableData.data.length > 0"
      />
    </template>
  </OnboardingPage>
</template>

<script>
import { PRODUCT_NAME, OFFLINE } from '@/consts'
import { getItemStatusFromInsight } from '@/dataplane'
import Kuma from '@/services/kuma'
import LoadingBox from '@/components/LoadingBox.vue'
import OnboardingNavigation from '@/views/Onboarding/components/OnboardingNavigation.vue'
import OnboardingHeading from '@/views/Onboarding/components/OnboardingHeading.vue'
import OnboardingPage from '@/views/Onboarding/components/OnboardingPage.vue'

export default {
  name: 'DataplanesOverview',
  components: {
    OnboardingNavigation,
    OnboardingHeading,
    OnboardingPage,
    LoadingBox,
  },
  metaInfo() {
    return {
      title: this.title,
    }
  },
  data() {
    return {
      productName: PRODUCT_NAME,
      tableHeaders: [
        { label: 'Mesh', key: 'mesh' },
        { label: 'Name', key: 'name' },
        { label: 'Status', key: 'status' },
      ],
      tableData: {
        total: 0,
        data: [],
      },
      timeout: null,
    }
  },

  computed: {
    title() {
      if (this.tableData.data.length) {
        return 'Success'
      }

      return 'Waiting for DPPs'
    },
    description() {
      if (this.tableData.data.length) {
        return 'The following data plane proxies (DPPs) are connected to the control plane:'
      }

      return null
    },
  },
  created() {
    this.getAllDataplanes()
  },
  beforeUnmount() {
    clearTimeout(this.timeout)
  },
  methods: {
    async getAllDataplanes() {
      let shouldRefetch = false
      const result = []

      try {
        const dataplanes = await Kuma.getAllDataplanes({ size: 10 })
        const items = dataplanes.items

        for (let i = 0; i < items.length; i++) {
          const { name, mesh } = items[i]

          const { status } = await Kuma.getDataplaneOverviewFromMesh({ mesh, name }).then((response) =>
            getItemStatusFromInsight(response.dataplaneInsight),
          )

          if (status === OFFLINE) {
            shouldRefetch = true
          }

          result.push({
            status,
            name,
            mesh,
          })
        }
      } catch (e) {
        console.error(e)
      }

      this.tableData.data = result
      this.tableData.total = this.tableData.data.length

      if (shouldRefetch) {
        this.timeout = setTimeout(() => {
          this.getAllDataplanes()
        }, 1000)
      }
    },
  },
}
</script>

<style lang="scss">
.onboarding-dataplane-table tbody tr {
  // hack to do not allow shadow overflow on top/bottom
  box-shadow: -3px 0 0 -1px var(--KTableBorder, var(--steel-200, #dae3f2)) !important;
}
</style>
