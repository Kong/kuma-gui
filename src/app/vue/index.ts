import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
