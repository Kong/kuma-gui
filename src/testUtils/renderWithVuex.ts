import { render, RenderOptions, RenderResult } from '@testing-library/vue'
import merge from 'lodash/merge'
import { StoreOptions } from 'vuex'
import Vue, { ComponentOptions } from 'vue'
import Store from '@/store'

function renderWithVuex(
  Component: typeof Vue | ComponentOptions<Vue>,
  options: RenderOptions<Vue> = { store: {}, routes: [] },
): RenderResult {
  const store: StoreOptions<any> = Store()

  // Render the component and merge the original store and the custom one
  // provided as a parameter. This way, we can alter some behaviors of the
  // initial implementation.

  return render(Component, { ...options, store: merge(store, options.store), routes: options.routes })
}

export default renderWithVuex
