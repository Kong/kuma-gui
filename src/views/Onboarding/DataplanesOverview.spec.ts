import renderWithVuex from '@/testUtils/renderWithVuex'
import { screen } from '@testing-library/vue'
import DataplanesOverview from './DataplanesOverview.vue'

describe('DataplanesOverview.vue', () => {
  // TODO before merge remove skip
  xit('renders snapshot', async () => {
    const { container } = renderWithVuex(DataplanesOverview, { routes: [] })

    expect(screen.getByTestId('loading')).toBeInTheDocument()
    expect(screen.getByText(/Waiting for DPPs/)).toBeInTheDocument()
    expect(screen.queryByText(/Next/)).not.toBeInTheDocument()

    await screen.findByText(/dataplane-test-456/)

    expect(container).toMatchSnapshot()
  })
})
