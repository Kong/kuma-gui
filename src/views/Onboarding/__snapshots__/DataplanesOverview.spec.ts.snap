// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataplanesOverview.vue renders snapshot 1`] = `
<div>
  <div>
    <div
      class="onboarding-container"
    >
      <div
        class="onboarding-container__header"
      >
        <div
          class="relative"
        >
          <h1
            class="onboarding-title"
          >
            
    Congratulations!
  
          </h1>
           
          <p
            class="text-center text-lg mt-3"
          >
            
    We have detected the following DPPs (dta plane proxies) connecting to this control plane
  
          </p>
           
          <div
            class="absolute right-0 top-0 mt-4 m-a flex items-center"
          >
            <button
              class="k-button skip-button small btn-link"
              data-v-38fa8ecf=""
              type="button"
            >
              
      Skip Setup
    
            </button>
          </div>
        </div>
      </div>
       
      <div
        class="onboarding-container__content"
      >
        <div
          class="w-full"
        >
          <div
            class="justify-center flex my-4"
          >
            <div>
              <p
                class="font-bold mb-4"
              >
                
          Found 3 DPPs, including:
        
              </p>
               
              <table
                class="k-table has-hover is-small side-border"
                data-v-bfeabd48=""
              >
                <thead
                  data-v-bfeabd48=""
                >
                  <tr
                    data-v-bfeabd48=""
                  >
                    <th
                      class=""
                      data-v-bfeabd48=""
                    >
                       Mesh 
                    </th>
                    <th
                      class=""
                      data-v-bfeabd48=""
                    >
                       Name 
                    </th>
                    <th
                      class=""
                      data-v-bfeabd48=""
                    >
                       Status 
                    </th>
                  </tr>
                </thead>
                <tbody
                  data-v-bfeabd48=""
                >
                  <tr
                    data-v-bfeabd48=""
                  >
                    <td
                      data-v-bfeabd48=""
                    >
                      default
                    </td>
                    <td
                      data-v-bfeabd48=""
                    >
                      dataplane-test-456
                    </td>
                    <td
                      data-v-bfeabd48=""
                    >
                      <div
                        class="entity-status"
                        data-v-bfeabd48=""
                      >
                        <span
                          class="entity-status__dot"
                          data-v-bfeabd48=""
                        />
                         
                        <span
                          class="entity-status__label"
                          data-v-bfeabd48=""
                        >
                          Online
                        </span>
                      </div>
                    </td>
                  </tr>
                  <tr
                    data-v-bfeabd48=""
                  >
                    <td
                      data-v-bfeabd48=""
                    >
                      default
                    </td>
                    <td
                      data-v-bfeabd48=""
                    >
                      ingress-dp-test-123
                    </td>
                    <td
                      data-v-bfeabd48=""
                    >
                      <div
                        class="entity-status"
                        data-v-bfeabd48=""
                      >
                        <span
                          class="entity-status__dot"
                          data-v-bfeabd48=""
                        />
                         
                        <span
                          class="entity-status__label"
                          data-v-bfeabd48=""
                        >
                          Online
                        </span>
                      </div>
                    </td>
                  </tr>
                  <tr
                    data-v-bfeabd48=""
                  >
                    <td
                      data-v-bfeabd48=""
                    >
                      default
                    </td>
                    <td
                      data-v-bfeabd48=""
                    >
                      no-subscriptions
                    </td>
                    <td
                      data-v-bfeabd48=""
                    >
                      <div
                        class="entity-status is-offline"
                        data-v-bfeabd48=""
                      >
                        <span
                          class="entity-status__dot"
                          data-v-bfeabd48=""
                        />
                         
                        <span
                          class="entity-status__label"
                          data-v-bfeabd48=""
                        >
                          Offline
                        </span>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="mt-6 flex items-center flex-col sm:flex-row justify-between"
      >
        <a
          aria-current="page"
          class="k-button navigation-button navigation-button--back router-link-exact-active router-link-active primary"
          data-v-38fa8ecf=""
          href="#/"
          type="button"
        >
          
    Back
  
        </a>
          
        <a
          aria-current="page"
          class="k-button navigation-button navigation-button--next router-link-exact-active router-link-active primary"
          data-v-38fa8ecf=""
          href="#/"
          type="button"
        >
          
    Next
  
        </a>
      </div>
    </div>
     
    <div
      class="background-image"
    />
  </div>
</div>
`;
